### ISSUE TITLE
Review request

### ISSUE THREAD

**Issue by konsumer** [konsumer/Qbs4QJS#1]
@bjorn I'd love your feedback on some new ideas I had. [Here is ScriptBinaryFile, as an example](https://github.com/konsumer/Qbs4QJS/blob/3009b261931532142eeaa49599dca69786cf63a9/scriptbinaryfile.h)

I started reading more closely the [QJSEngine docs](https://doc.qt.io/qt-5/qjsengine.html) and wanted to come up with an implementation that more closely matches the documented way of exposing things, using very little other code & what Qt5 already has built-in. It might change things a bit, but I think the final product will be simpler & easier to maintain, and more closely match how Qbs works.

[Here is a usage example](https://github.com/konsumer/Qbs4QJS/blob/3009b261931532142eeaa49599dca69786cf63a9/test.js).

Does this make sense?

I think it still needs some work, for example if I call `b1.close()` in the [example](https://github.com/konsumer/Qbs4QJS/blob/3009b261931532142eeaa49599dca69786cf63a9/test.js) it silently fails, without an error message.

**Comment by konsumer** [comment]
Hmm, I might need to rethink this approach.

If I run this code:

```js
const b1 = new BinaryFile('test.bin', BinaryFile.ReadOnly)
console.log(b1.filePath())
try {
  console.log(b1.read(b1.size()))
} catch (e) {
  console.error('Error reading file.', e.message)
}
```

and `'test.bin` doesn't exit, i get this:

```
qml: /home/konsumer/Documents/otherdev/Qbs4QJS/test.bin
QIODevice::read (QFile, "test.bin"): device not open
qml: 
qml: Qbs4QJS::BinaryFile(0x55b36f82feb0)
```

which didn't use my try/catch.

**Comment by konsumer** [comment]
I'm going to ask around on reddit and whatnot. I think I am missing some basic understanding of how it fits together, that I can't workout form the examples in the docs.

**Comment by bjorn** [comment]
> which didn't use my try/catch.

Why would it? You've just written
```c++
    Q_INVOKABLE QByteArray read(qint64 size) {
        return m_file->read(size);
    }
```
Which does not raise a JS exception in case of error.

What is the main difference you're after when you compare this `ScriptBinaryFile` implementation with the one in the Tiled repository?

**Comment by konsumer** [comment]
> Why would it? You've just written

The thing I am pointing out is that it silently fails (and exits status 1) and I don't totally know why. It seems to me like if something makes it exit status 1, that the JS engine is saying "that didn't work" and would therefore hit my my try catch, or the unhanded exception code. That doesn't seem to to be the case.  The `read` is throwing an error because the file isn't open. I would like to throw my own error in constructor, when the open fails, but I have no access to `jsEgnine`.

As I stated on the other issue, I think I might need to drop this idea of putting it in a class. An instance object is much simpler, even if it'll end up making a bunch of temporary `QFileInfo` and `QFile` objects. I looked around, and this seems to be a common issue with porting `QScriptEngine` code to `QJSEngine`: there is no reference to the calling engine, so throwing errors in JS requires some trickery. I want the library to be useful for multiple situations, and the idea of using a global or a singleton seems a bit counter to that.

**Comment by bjorn** [comment]
> The thing I am pointing out is that it silently fails (and exits status 1) and I don't totally know why. It seems to me like if something makes it exit status 1, that the JS engine is saying "that didn't work" and would therefore hit my my try catch, or the unhandled exception code.

What do you mean with "exit status 1"? There's just a warning printed to standard error and an empty `QByteArray` is returned to the JS engine. The JS engine has no way of detecting that an error happened.

> I would like to throw my own error in constructor, when the open fails, but I have no access to `jsEngine`.

Right, that's probably not possible. In the constructor your object is not fully constructed yet so you can't use [`qjsEngine(this)`](https://doc.qt.io/qt-5/qjsengine.html#qjsEnginex) to get the engine and throw an error. In the `read` function it would be an option though. This means you could solve the issue by not doing anything that raises an error in the constructor, and instead add a `BinaryFile.open(mode)` method that needs to be called separately.

In Tiled I cheated a bit by accessing the script engine through the `ScriptManager`, which an independent library can't do. It did show that as long as you can get a pointer to the JS engine, throwing an error in the constructor does appear to be supported...

Regarding global or singleton, this is not suitable for objects such as the `BinaryFile` class. This object has a state which enables reading and writing in increments. A global helper function to read or write a file would generally have to read or write the entire file at once, or the read function would need to take an offset + size and the write function an option to open for append, but opening and closing the file all the time would get rather slow.

**Comment by konsumer** [comment]
> What do you mean with "exit status 1"?

In bash. `npm test` says "the test failed" if it exits status 1, or you can also do `qbs run -- test.js || echo "failed"`. It will have exit status 0, if everything goes right (the file exists, in this case.) 

> In the constructor your object is not fully constructed yet so you can't use `qjsEngine(this)` to get the engine and throw an error.

Ah! Maybe I compromise on making it Qbs-like in a different way. If I made a seperate `open` function it might be better, anyway. I can make the constructor just not throw. I didn't know I could do this to get the the engine.

> In Tiled I cheated a bit by accessing the script engine through the `ScriptManager`

> Regarding global or singleton, this is not suitable for objects such as the BinaryFile class. 

This is what I mean. `ScriptManager::instance()` forms a singleton. I wanted to not have a singleton or global, so it could be attached to multiple javascript-engines, but I was having trouble working it out. I will experiment with a new `open` method.

**Comment by konsumer** [comment]
Hmm, I tried that in [7381ab9](https://github.com/konsumer/Qbs4QJS/commit/7381ab9cfff9f8369624319a757cd3fa8a4fcf6a), and it's still not throwing how I would expect, but it doesn't have any errors. Maybe `open` is happy with the file not existing, but `read` is not, so I need to throw an error on read, too.

**Comment by konsumer** [comment]
Ok! I think I got it. Thanks for your help @bjorn. Does [this](https://github.com/konsumer/Qbs4QJS/blob/d871a49d53ead803521505d4458027eb6548212c/binaryfile.h) seems like an ok way to go about this? If so, I can use similar approach for other classes.

**Comment by bjorn** [comment]
> I just mean the [CLI exit status](https://www.cyberciti.biz/faq/bash-get-exit-code-of-command/). `npm test` says "the test failed" if it exits status 1, or you can also do `qbs run -- test.js || echo "failed"`. It will have exit status 0, if everything goes right (the file exists, in this case.)

Right, but I fail to see the connection with this exit status and the try/catch in your script above.

But in any case, indeed you almost got it working fine I think. I just noticed that your `openCheck` function is not returning whether the file is there, so after throwing an error in the JS engine, it will return and then you'll get a crash due to using the null `m_file` member.

**Comment by konsumer** [comment]
> Right, but I fail to see the connection with this exit status and the try/catch in your script above.

Hmm, maybe I'm not describing clearly. I just mean "it calls the general `exit=1` code in main, but no other error-handling is happening, like the catch in JS." It is working well, now, though, and JS can handle errors.

> I just noticed that your openCheck function is not returning whether the file is there, so after throwing an error in the JS engine, it will return and then you'll get a crash due to using the null m_file member.

Ah, I though the throw would stop operation for some reason, like in JS, but that makes sense. I will add it.

**Comment by konsumer** [comment]
@bjorn ok, I think I got all your suggestions in. I will work on the other classes in a bit. Thanks!

**Comment by konsumer** [comment]
@bjorn I think I got all the main classes in there, and the basic style/structure for the lib. I need to write a lot more tests, but the basic code should be in place. If you have any concerns about how any of it's setup, I'm very open to suggestions.
